### PR TITLE
Delay webhook trigger events to avoid GitLab cache

### PR DESCRIPTION
**Pull request by cpt1gl0**
As multi-branch requests are being cached by GitLab API, this MR introduces a delay for webhook triggers to wait for GitLab cache to expire before triggering SCM events. In that way we can avoid fetching outdated branches from GitLab, which would result in builds not being triggered due to undetected changes.

Solves #144.

### CODE REVIEW

**Comment by cpt1gl0** [comment]
@EugeneLesnov:
Hi Eugene! May I once more ask for a review?

**Review by EugeneLesnov — APPROVED** [review]

**Comment by EugeneLesnov** [comment]
Looks good. Anyway, let's wait for some feedback.

**Comment by cpt1gl0** [comment]
@EugeneLesnov Changed my fix since it would not have worked with new branches being created. The new strategy is to wait for GitLab API cache to expire and trigger the SCM event after expiry. I am kindly asking for another review...

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java
Comment (EugeneLesnov) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java:18]: It might be a good idea to make this parameter configurable..

Comment (cpt1gl0) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java:18]: I was thinking about it, but then decided not to have it configurable because also for GitLab the cache timeout is hard-coded to be 30 seconds. As no other value than the GitLab hard-coded timeout would make sense here, I was thinking there would possibly be less confusion if users could not configure it.

Comment (EugeneLesnov) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java:18]: Oh, I see. Then let's wait for merging :)

Comment (cpt1gl0) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java:18]: Okay, let's wait then. Thanks for reviewing!

Comment (LinuxSuRen) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java:17]: Considering the [discussion from Gitlab team](https://gitlab.com/gitlab-org/gitlab/-/issues/333548), it seems like a temporary solution.

Secondly, the delay issue only affects parts of the Gitlab versions. Let assume there is one user who has an old Gitlab, then his jobs cannot be triggered in time.

So, I prefer to have two changes here:

* Make the timeout value to be configurable in case of Gitlab changes it
  * Make it be an advanced configuration item, might be easier for some users. Normally, people don't need to change it.
* Don't have a timeout value if it's an old Gitlab. You can get the version from Gitlab API, then cache it.

**Review by LinuxSuRen — CHANGES_REQUESTED** [review]
hi @EugeneLesnov and @cpt1gl0 , thanks for your time. I left some comments. Please take a look at it.

**Comment by cpt1gl0** [comment]
@LinuxSuRen I implemented the changes you suggested so that now the delay can be manually overridden and the default delay is only used for GitLab versions >= 13.12.0. Unfortunately there is an issue when retrieving the GitLab API version from the web hook listener, as the listener is not allowed to do that and version retrieval fails with:
`hudson.security.AccessDeniedException3: anonymous is missing the Job/Build permission`

Can you point me in the right direction where to implement GitLab version retrieval? You were talking about caching API versions in your original comment.

**Comment by cpt1gl0** [comment]
@EugeneLesnov Hey Eugene! Do you know a good place in the code where to retrieve GitLab API versions and cache them for all servers?

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java
Comment (cpt1gl0) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java]: - Made the timeout configurable
- I tried to implement GitLab API version check to disable the trigger delay for old versions but failed because I could not find a suitable place to retrieve versions from GitLab API and cache them. Retrieving API version from WebHookListener gives an exception because WebHookListener is not allowed to access GitLab credentials.

**Comment by cpt1gl0** [comment]
@LinuxSuRen 
- As you suggested I made the trigger delay configurable through an advanced config item.
- I did not disable the delay for older GitLab versions because I could not find a suitable place in the code to retrieve and cache GitLab API versions for all servers. Retrieving API version directly from WebHookListener is not possible because GitLab credentials can not be accessed from within WebHookListener.

If you can live with the delay not being disabled for older GitLab versions the PR is ready to be merged now!

**Comment by cpt1gl0** [comment]
@LinuxSuRen Hi! Could you please merge this PR if you are fine with the changes?

**Review by cpt1gl0 — COMMENTED** [review]
All done.

**Review by LinuxSuRen — APPROVED** [review]

**Pull request merged**
